hub/core/chunk_engine/generator.py

```diff
@@ -0,0 +1,87 @@
+import numpy as np
+from typing import Generator, Tuple
+
+from hub.core.chunk_engine.exceptions import ChunkGeneratorError
+
+
+def generate_chunks(
+    content_bytes: bytes,
+    chunk_size: int,
+    last_chunk_num_bytes: int = None,
+) -> Generator[Tuple[bytearray, int], None, None]:
+    """
+    Generator function that chunks bytes.
+
+    Chunking is the process of taking the input `content_bytes` & breaking it up into a sequence of smaller bytes called "chunks".
+    The sizes of each chunk are <= `chunk_size`.
+
+    Example:
+        content_bytes = b"1234567890123"
+        chunk_size = 4
+        yields:
+            b"1234", 1
+            b"5678", 2
+            b"9012", 3
+            b"3", 4
+
+    Args:
+        content_bytes (bytes): Bytes object with the data to be chunked.
+        chunk_size (int): Each individual chunk will be assigned this many bytes maximum.
+        last_chunk_bytes (int, optional): If chunks were created already, `last_chunk_bytes`
+            should be set to the length of the last chunk created. This is so the generator's
+            first output will be enough bytes to fill that chunk up to `chunk_size`.
+
+    Yields:
+        bytearray: Chunk of the `content_bytes`. Will have length on the interval (0, `chunk_size`].
+        int: Relative index of the yielded chunk (bytes) to previously created chunks (if any).
```

Comment:
What's the point of yielding the index? And why not zero-indexed?

Surely it's sufficient to just be a simple generator and then allow the caller to use `enumerate` if they want

Reply:
The index tells you whether or not the output chunk should be appended to the previous.

Reply:
This is useful for building our index map & keeping it well synchronized.

Reply:
oh, but i guess the handler that calls this function can just get that themselves. you're right i will remove this.

Reply:
I see now what you're doing with the indexing. I still think it can be done better, but I suppose this will work.